indico/modules/events/timetable/util.py

```diff
@@ -346,6 +349,112 @@ def get_time_changes_notifications(changes, tzinfo, entry=None):
+    children_contrib_strategy.joinedload('person_links')
+    children_contrib_strategy.joinedload('subcontributions')
+    children_contrib_strategy.joinedload('references')
+    children_contrib_strategy.joinedload('own_room')
+
+    children_subcontrib_strategy = children_contrib_strategy.joinedload('subcontributions')
+    children_subcontrib_strategy.joinedload('person_links')
+    children_subcontrib_strategy.joinedload('references')
+
+    contrib_strategy = joinedload('contribution')
+    contrib_strategy.joinedload('person_links')
+    contrib_strategy.joinedload('references')
+
+    subcontrib_strategy = contrib_strategy.joinedload('subcontributions')
+    subcontrib_strategy.joinedload('person_links')
+    subcontrib_strategy.joinedload('references')
+
+    if include_notes:
+        children_contrib_strategy.joinedload('note')
+        contrib_strategy.joinedload('note')
+        subcontrib_strategy.joinedload('note')
+
+    # try to minimize the number of DB queries
+    options = [contrib_strategy,
+               children_strategy,
+               joinedload('session_block').joinedload('person_links'),
+               joinedload('session_block').joinedload('own_room'),
+               joinedload('break_')]
+
+    entries = []
+
+    for entry in event.timetable_entries.filter_by(parent=None).options(*options):
+        if show_date != 'all' and entry.start_dt.astimezone(event_tz).date().isoformat() != show_date:
+            continue
+        if (entry.type == TimetableEntryType.CONTRIBUTION and
+                (detail_level not in ('contribution', 'all') or show_session != 'all')):
+            continue
+        elif (entry.type == TimetableEntryType.SESSION_BLOCK and show_session != 'all' and
+                str(entry.object.session.friendly_id) != show_session):
+            continue
+
+        if entry.type == TimetableEntryType.BREAK:
+            entries.append(entry)
+        elif entry.object.can_access(session.user):
+            entries.append(entry)
+
+    entries.sort(key=attrgetter('end_dt'), reverse=True)
+    entries.sort(key=lambda entry: (entry.start_dt, *_entry_title_key(entry)))
+
+    return entries
+
+
+def get_nested_timetable_location_conditions(entries):
+    show_siblings_location = False
+    show_children_location = {}
+
+    for entry in entries:
+        show_children_location[entry.id] = not all(
+            child.object.inherit_location for child in entry.children
+        )
+
+        if (
+            not show_siblings_location and
+            # the object itself does not inherit
+            (not entry.object.inherit_location or
+            # the object is a session block and inherits from a session with a custom location
+            (entry.type == TimetableEntryType.SESSION_BLOCK and
+                entry.object.inherit_location and
+                not entry.object.session.inherit_location))
+        ):
+            show_siblings_location = True
```

Comment:
The formatting from ruff is also not SUPER nice (I don't like those operators in their own line), but I think it's clearer since in your version it looks like 3 conditions, even though the "third" one is nested.

```suggestion
        if (
            not show_siblings_location
            and
            # the object itself does not inherit
            (
                not entry.object.inherit_location
                or
                # the object is a session block and inherits from a session with a custom location
                (
                    entry.type == TimetableEntryType.SESSION_BLOCK
                    and entry.object.inherit_location
                    and not entry.object.session.inherit_location
                )
            )
        ):
            show_siblings_location = True
```

Reply:
However, I wonder if this would not be more readable, even though it means iterating over `entries` twice (which is likely negligible from a performance point of view):

```py
def get_nested_timetable_location_conditions(entries):
    show_siblings_location = False
    show_children_location = {}

    show_siblings_location = any(
        not entry.object.inherit_location or (
            # the object is a session block and inherits from a session with a custom location
            entry.type == TimetableEntryType.SESSION_BLOCK
            and entry.object.inherit_location
            and not entry.object.session.inherit_location
        )
        for entry in entries
    )

    show_children_location = {
        entry.id: not all(child.object.inherit_location for child in entry.children)
        for entry in entries
    }

    return show_siblings_location, show_children_location
```

(untested of course)

Reply:
The main bottleneck is fetching the entries from the database so iterating twice over our entries should indeed not be too bad, but I'll check to be sure

Reply:
Works and performance is fine even with bigger events